README


This archive contains the distribution Farabi,
version 0.46:

  Modern Perl IDE

This software is copyright (c) 2014 by Ahmad M. Zawawi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.016.



Changes
Changes for Farabi: The Modern Perl IDE

0.46      2014-05-17 15:09:03+03:00 Asia/Baghdad

    * Please try Farabi and let me know what you like and what loathe :)
    * SECURITY WARNING: Please do not think of deploying this release on the internet. 
      You have been warned!

[ADDED]
    - Added 'minil test' and 'dzil test' to build menu for an initial seed for the Minilla and Dist::Zilla runtime project support.
    - Alt-N: close file, Alt-O: open file, Alt-S: save file, Alt-W: close file, Alt-L: goto line
    - Added 'ack' runtime support. Please press Ctrl-Alt-F to find selected text in current project folder
    - Toggle full screen editor editing by F11 or F10. Some browser/OS combinations block some keyboard shortcuts.
    - Added Markdown auto preview support. Pod tab is now called "Preview"
    - Farabi can now auto preview POD (if the preview tab is visible) in 
      cpan, metacpan, github CSS themes (Courtesy of "POD Web view" by Michał Wojciechowski).
      Please use the "POD style" option in Tools/Options to change it.

[CHANGED]
    - Upgrade CodeMirror to 4.1
    - Less installation dependencies; more runtime dependencies
    - Added Spellunker runtime support now displays errors in problems. 
    - Added App::midgen runtime support to generate Makefile.PL et all requires/recommends
      and display results in search tab.
    - Documentation updates
    - Problems tab is the first tab now
    - Fix codemirror mode error on plain text files
    - "Changes/Git Diff" tab is now updated on an open/save file or mouse click when it is visible.
    - Farabi is now using simple AJAX request instead of buggy websocket support
    - Ping HTTP message is now sent every 5 seconds
    - Run menu is now called Build
    - Use runtime Code::CutNPaste support to find duplicate code
    - Disable incomplete features like new project, help documentation/search
    - git diff, and jshint are now enabled if you have their commands in your $PATH. 
      Please note that JSHint is still executed in the client-side.
    - Options dialog is more beautiful or at least to my eyes :)
    - Spellunker::Pod runtime support is here for checking your POD spelling.
    - Perl::Strip (to minify your Perl script) runtime support
    - Added Perl::Tidy and Perl::Critic runtime support.
    - TODO is now TODO.md
    - Syntax/preview actions are triggered on editor focus and change
    - Remove buggy auto bracket codemirror support.
    - Remove Perl 6 (rakudo) and parrot support. Farabi will focus on Perl script and web-based development.
    - Update theme list to contain CodeMirror's full list
    - Debug is now "Documentation" tab. Currently it does not do anything useful.
    - Diff tab is now called "Changes"
    - POD tab is now called "Preview".
    - If preview tab is visible, POD rendering on editor focus/change is active
    - Remove plugin manager skeleton implementation. It sucked. Farabi is not Padre :)
    - Show keyboard shortcuts in the menu and help dialog
    - Remove selected text workaround patch. CodeMirror 4.x already include it in its themes
    - Remove repetitive and slow dialog fade animation effect
    - Auto Git differences when Git tab is open
    - Remove incomplete Debig::Client; Debug::Client test suite stalled on some Travis CI build runs
      Please note that is not truly fullscreen editing. It is an experiment in progress.

    [FIXED]
    - Utilize the theme specific active line and gutter from CodeMirror mode CSS

0.45      2014-05-13 14:01:12+03:00 Asia/Jerusalem

    [ADDED]
    - TravisCI GitHub integration is now working
    - Added initial Splunker Spell checking (IN PROGRESS)
    - Various package dependencies on Minilla, Debug::Client and Reply (IN PROGRESS)

    [CHANGED]
    - Upgrade CodeMirror to 3.12

    [FIXED]
    - Merge pull #19 "pod fix" dsteinbrunner++
    - Merge pull #20 "Mojolicious 4.91 no longer has 'secret' method" chorny++
    - Various Documentation fixes (shlomif++)
    - Require Parse::ErrorString::Perl 0.18 to get rid of the perl 5.18+ warning line noise

0.44      2013-04-08 11:21:16 Asia/Baghdad

    [CHANGED]
    - Upgrade JSHint to 1.1.0
    - Upgrade Bootstrap to 2.3.1
    - Upgrade CodeMirror to 3.11

0.43      2013-04-07 15:01:24 Asia/Baghdad

    [ADDED]
    - Added "Perl Strip" action in the tool menu to reduce script size by stripping
      whitespace, comments, pod etc using Perl::Strip
    - Add more documentation about the output tab
    - No need for exit code at the end of a diff
    - Added Diff tab along with a 'Git Diff' action in the tools menu
    - Added various dummy debug actions

    [FIXED]
    - Fix JSHint version in About dialog
    - Require mojolicious 3.89 to get the fix for dual life autodie under morbo
    - Fix editor refresh bug in tabs.

    [CHANGED]
    - Sort prerequisties in dist.ini
    - Enabled dummy 'New Project' action

0.42      2013-03-04 16:44:55 Asia/Baghdad

    [ADDED]
    - Added the option to toggle the visibility of tabs and spaces (default: on)
    - Added an input tab that allows input text to be sent as STDIN to the script when run.

    [FIXED]
    - Fix regression where editor is sometimes not visible when a file is opened.
      This regression happened from the 3.1 CodeMirror upgrade.
    - Fixed various JSHint warnings.

    [CHANGED]
    - Upgrade to JSHint 1.0.0 stable

0.41      2013-02-25 17:50:34 Asia/Baghdad

    [FIXED]
    - Fix Test::MinimumVersion dependency typo

0.40      2013-02-25 17:43:30 Asia/Baghdad

    [ADDED]
    - Added auto close brackets feature
    - Added t/02-minimum-per.t to make sure post 5.10.1-isms do not creep into the code
    - Added File::HomeDir 1.00 and Path::Tiny 0.013 to dependencies
    - Add a disabled "New Project" action/dialog. This will be enabled once it is ready
    - Added [Test::Compile] to dist.ini to make sure all files are syntax checked

    [CHANGED]
    - Upgrade to CodeMirror v3.1.

lib/Farabi.pm
package Farabi;

use Mojo::Base 'Mojolicious';
use Path::Tiny;

# ABSTRACT: Modern Perl IDE
our $VERSION = '0.46'; # VERSION


# Application SQLite database and projects are stored in this directory
has 'home_dir';

# Projects are stored in this directory
has 'projects_dir';

# The database name and location
has 'db_name';

sub startup {
	my $app = shift;

	# Change secret passphrase that is used for signed cookies
	$app->secrets(['Hulk, Smash!']);

	# Use content from directories under lib/Farabi/files
	$app->home->parse( path( path(__FILE__)->dirname, 'Farabi' ) );
	$app->static->paths->[0]   = $app->home->rel_dir('files/public');
	$app->renderer->paths->[0] = $app->home->rel_dir('files/templates');

	# Define routes
	my $route = $app->routes;
	$route->get('/')->to('editor#default');
	$route->post("/syntax_check")->to('editor#syntax_check');
	$route->post('/pod2html')->to('editor#pod2html');
	$route->post("/md2html")->to('editor#md2html');
	$route->post("/perl_critic")->to('editor#perl_critic');
	$route->post("/perl_tidy")->to('editor#perl_tidy');
	$route->post("/perl_strip")->to('editor#perl_strip');
	$route->post("/spellunker")->to('editor#spellunker');
	$route->post("/code_cutnpaste")->to('editor#code_cutnpaste');
	$route->post("/git_diff")->to('editor#git_diff');
	$route->post("/open_file")->to('editor#open_file');
	$route->post("/save_file")->to('editor#save_file');
	$route->post("/find_file")->to('editor#find_file');
	$route->post("/find_action")->to('editor#find_action');
	$route->post("/run_perl")->to('editor#run_perl');
	$route->post("/run_perlbrew_exec")->to('editor#run_perlbrew_exec');
	$route->post("/dump_ppi_tree")->to('editor#dump_ppi_tree');
	$route->post("/repl_eval")->to('editor#repl_eval');
	$route->post("/ping")->to('editor#ping');
	$route->post("/ack")->to('editor#ack');
	$route->post("/midgen")->to('editor#midgen');
	$route->post("/minil_test")->to('editor#minil_test');
	$route->post("/dzil_test")->to('editor#dzil_test');

	eval { $app->_setup_dirs };
	if ($@) {
		die "Failure to create \$HOME/.farabi directory structure, reason: $@";
	}

	# The database name
	$app->db_name(path($app->home_dir, 'farabi.db'));

	# Setup the Farabi database
	eval { $app->_setup_database };
	if ($@) {
		warn "Database not setup, reason: $@";
	}
}

sub support_can_be_enabled {
	my $app = shift;
	my $module = shift;

	my %REQUIRED_VERSION = (
		'Perl::Critic' => '1.118',
		'Perl::Tidy' => '20121207',
		'Perl::Strip' => '1.1',
		'Spellunker'  => '0.0.17',
		'Code::CutNPaste' => '0.04',
		'App::Midgen' => '0.32',
		'Minilla' => '0.4.6',
		'Dist::Zilla' => '5.016',
	);

	my $version = $REQUIRED_VERSION{$module};
	return 0 unless defined $version;

	eval qq{use $module $version;};
	if($@) {
		$app->log->warn("$module support is disabled. Please install $module $version or later.");
		return 0;
	} else {
		$app->log->info("$module support is enabled");
		return 1;
	}
}

#
# Create the following directory structure:
# .farabi
# .farabi/projects
#
sub _setup_dirs {
	my $app = shift;

	require File::HomeDir;

	$app->home_dir( path( File::HomeDir->home, ".farabi" ) );
	$app->projects_dir( path( $app->home_dir, "projects" ) );
	$app->projects_dir->mkpath;
}

# Setup the Farabi database
sub _setup_database {
	my $app = shift;

	# Connect and create the Farabi SQLite database if not found
	require DBIx::Simple;
	my $db_name = $app->db_name;
	my $db = DBIx::Simple->connect("dbi:SQLite:dbname=$db_name");

	# Create tables if they do not exist
	$db->query(<<SQL);
CREATE TABLE IF NOT EXISTS recent_list (
	id        INTEGER PRIMARY KEY AUTOINCREMENT, 
	name      TEXT,
	type      TEXT,
	last_used TEXT
)
SQL

	# Disconnect from database
	$db->disconnect;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Farabi - Modern Perl IDE

=head1 VERSION

version 0.46

=head1 SYNOPSIS

  # Run on the default port 4040
  $ farabi
  
  # Run it on port 5050
  $ farabi --port 5050

=head1 DESCRIPTION

This is a modern web-based Perl IDE that runs inside your favorite browser.

Please run the following command and then open http://127.0.0.1:4040 in your browser:

  farabi

=head1 SECURITY WARNING

B<Farabi is an experiment in progress>. It is a web-based user interface with a backend Perl web server.
Please B<DO NOT> serve it on the Internet unless you jail it in an isolated uber-secure 
environment that has proper CPU and I/O limits and non-root access.

You have been warned, young padawan :)

=head1 FEATURES

=over

=item Open File(s)

The dialog provides partial filename search inside the directory where Farabi was started.
Matched single or multiple file selections can then be opened in one batch.

B<WARNING:> Please do not start farabi in a folder with too many files like your home directory
because this feature's performance will eventually suffer.

=back

=head1 METHODS

=head1 support_can_be_enabled

Returns 1 when a required C<module> with a specific version is found otherwise returns 0.

It can be used in the future to toggle feature XYZ runtime support

=head1 TECHNOLOGIES USED

=over

=item *

L<Mojolicious|http://mojolicio.us> - A next generation web framework for the Perl programming language

=item *

L<jQuery|http://jquery.com/> - A new kind of JavaScript Library

=item *

L<JSHint|http://jshint.com/> - A JavaScript Code Quality Tool

=item *

L<Bootstrap|http://twitter.github.com/bootstrap> - Sleek, intuitive, and powerful front-end framework for faster and easier web development

=item *

L<CodeMirror|http://codemirror.net> - In-browser code editing made bearable

=item *

L<Perlito|http://perlito.org/> - Runtime for "Perlito" Perl5-in-Javascript

=back

=head1 SEE ALSO

L<EPIC|http://www.epic-ide.org/>, L<Kephra>, L<Padre>, L<TryPerl|http://tryperl.com/>

=head1 HISTORY

The idea started back in March 2012 as a fork of L<Padre>. I wanted to dump L<Wx> for the browser. 
The first version was in 11th April as L<Mojolicious::Plugin::Pedro>. It used the ACE Javascript
editor and jQuery UI. Then i hibernated for a while to play games :) Later I heard about L<Galileo>.
It basically used the same idea, mojolicious backend, browser for the frontend. So I stopped 
playing games and rolled my sleeves to focus on Pedro.

Later I discovered Pedro was not a good name for my project. So I chose Farabi for
L<Al-Farabi|http://en.wikipedia.org/wiki/Al-Farabi> who was a renowned scientist and philosopher
of the Islamic Golden Age. He was also a cosmologist, logician,and musician.

=head1 SUPPORT

If you find a bug, please report it in:

L<https://github.com/azawawi/farabi/issues>

If you find this module useful, please rate it in:

L<http://cpanratings.perl.org/d/Farabi>

=head1 AUTHORS

Ahmad M. Zawawi E<lt>ahmad.zawawi@gmail.comE<gt>

=head1 CONTRIBUTORS

Kevin Dawson E<lt>bowtie@cpan.orgE<gt>

=head1 AUTHOR

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Ahmad M. Zawawi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Farabi/Controller.pm
package Farabi::Controller;
use Mojo::Base 'Mojolicious::Controller';

our $VERSION = '0.02';

use Config;

# Taken from Padre::Plugin::PerlCritic
sub perl_critic {
	my $self = shift;
	my $source = $self->param('source');

	# Check for problems
	unless ( defined $source ) {
		$self->render( json => [] );
		return;
	}

	# Hand off to Perl::Critic
    require Perl::Critic;
    my @violations = Perl::Critic->new->critique( \$source );

	my @results;
	for my $violation (@violations) {
		push @results, {
			policy      => $violation->policy,
			line_number => $violation->line_number,
			description => $violation->description,
			explanation => $violation->explanation,
			diagnostics => $violation->diagnostics,
		};
	}

 	$self->render( json => \@results );
}

# Taken from Padre::Plugin::PerlTidy
# TODO document it in 'SEE ALSO' POD section
sub perl_tidy {
	my $self   = shift;
	my $source = $self->param('source');
	
	my %result = (
		'error'  => '',
		'source' => '',
	);

	# Check for problems
	unless ( defined $source ) {
		$self->render( json => \%result );
		return;
	}

	my $destination = undef;
	my $errorfile   = undef;
	my %tidyargs    = (
		argv        => \'-nse -nst',
		source      => \$source,
		destination => \$destination,
		errorfile   => \$errorfile,
	);

	# TODO: suppress the senseless warning from PerlTidy
	eval {
		require Perl::Tidy;
		Perl::Tidy::perltidy(%tidyargs);
	};

	if ($@) {
		$result{error} = "PerlTidy Error:\n" . $@;
	}

	if ( defined $errorfile ) {
		$result{error} .= "\n$errorfile\n";
	}

	$result{source} = $destination;

	return $self->render( json => \%result );;
}

sub help_search {
	my $self = shift;
	my $topic = $self->param('topic') // '';

	require File::Spec;
	my $pod_path = File::Spec->catfile( $Config{archlibexp}, 'pods' );
	my $pod_index_filename = 'index.txt';
	unless ( -f $pod_index_filename ) {

		# Create an index
		say "Creating POD index";
		require Pod::Index::Builder;
		my $p     = Pod::Index::Builder->new;
		require File::Glob;
		my @files = File::Glob::glob(File::Spec->catfile($pod_path, '*.pod'));
		my $t0    = time;
		for my $file (@files) {
			say "Parsing $file";
			$p->parse_from_file($file);
		}
		say "Job took " . ( time - $t0 ) . " seconds";
		$p->print_index($pod_index_filename);
	}

	# Search for a keyword in the file-based index
	require Pod::Index::Search;
	my $q = Pod::Index::Search->new(
		filename => $pod_index_filename,
		filemap  => sub {
			my $podname = shift;
			if($podname =~ /^.+::(.+?)$/) {
				$podname = File::Spec->catfile($pod_path, "$1.pod");
			}
			return $podname;
		}
	);

	my @results = $q->search($topic);
	my @help_results;
	for my $r (@results) {
		next if $r->podname =~ /perltoc/;
		my $podname = $r->podname;
		$podname =~ s/^.+::(.+)$/$1/;
		push @help_results, {
			'podname' => $podname,
			'context' => $r->context,
			'html'    => _pod2html($r->pod),
		};
	}

	$self->render( json => \@help_results );
}

sub _pod2html {
	my $pod = shift;

	require Pod::Simple::XHTML;
	my $psx = Pod::Simple::XHTML->new;
	$psx->no_errata_section(1);
	$psx->no_whining(1);
	$psx->output_string(\my $html);
	$psx->parse_string_document($pod);

	return $html;
}

sub default {
	my $self = shift;

	# Render template "controller/default.html.ep"
	$self->render;
}

1;


lib/Farabi/files/public/farabi.js
/**
 * Starts Farabi... :)
 */
function startFarabi(editorId) {

	// Cache dialog jQuery references for later use
	var $perl_critic_dialog = $('#perl_critic_dialog');
	var $about_dialog = $('#about_dialog');
	var $help_dialog = $("#help_dialog");

	$("#theme_selector").change(function() {
		var $selectedTheme = $(":selected", this);
		var theme = $selectedTheme.val();

		$("head").append("<link>");
		var css = $("head").children(":last");
		css.attr({
			rel:  "stylesheet",
			type: "text/css",
			href: "assets/codemirror/theme/" + theme + ".css"
		});

		editor.setOption("theme", theme);

		humane.log("Mode changed to " + $selectedTheme.html());
	});

	function displayHelp(cm) {
		var selection = cm.getSelection();
		if(selection) {
			_displayHelp(selection, true);
		} else {
			// Search for token under the cursor
			var token = cm.getTokenAt(cm.getCursor());
			if(token.string) {
				_displayHelp($.trim(token.string), true);
			} else {
				_displayHelp('', true);
			}
		}
	}

	function _displayHelp(topic, bShowDialog) {
		$.post('/help_search', {"topic": topic}, function(results) {
			if(results.length > 0) {
				humane.log("Found " + results.length + " help results for '" + topic +"'!");
				$(".topic").val(topic);
				var html = '';
				for(var i = 0; i < results.length; i++) {
					html += '<option value="' + i + '">' + results[i].podname + "  (" + results[i].context + ")" +'</option>';
				}

				$(".results")
					.html(html)
					.change(function() {
						var index = $(':selected', this).val();
						$(".preview").html(results[index].html);
					}).change().show();
			} else {
				humane.log("No help found for '" + topic + "'");

				$(".topic").val(topic);
				$(".results").html('').hide();
				$(".preview").html('<span class="text-error">No results found!</span>');
			}
			if(bShowDialog) {
				$('a[href="#learn-tab"]').tab('show');
				$(".topic").select().focus();
			}
		});
	}

	function changeMode(cm, modeFile, mode) {
		if(typeof mode == 'undefined') {
			mode = modeFile;
		}
		CodeMirror.modeURL = "assets/codemirror/mode/%N.js";
		cm.setOption("mode", mode);
		CodeMirror.autoLoadMode(cm, modeFile);	
	}

	function plural(number) {
		if(number == 1) {
			return 'st';
		} else if(number == 2) {
			return 'nd';
		} else {
			return 'th';
		}
	}

	function showEditorStats(cm) {
		var cursor = cm.getCursor();
		var selection = cm.getSelection();
		var line_number = cursor.line + 1;
		var column_number = cursor.ch + 1;
		$('#editor_stats').html(
			'<strong>' + line_number + '</strong>' + plural(line_number) + ' line' +
			', <strong>' + column_number + '</strong>' + plural(column_number) + ' column' +
			', <strong>' + cm.lineCount() + '</strong>&nbsp;lines' +
			(selection ?  ', <strong>' + selection.length +'</strong>&nbsp;selected characters' : '')  +
			', <strong>' + cm.getValue().length + '</strong>&nbsp;characters' +
			', <strong>' + cm.lineCount() + '</strong>&nbsp;Lines&nbsp;&nbsp;'
		);
	}

	$("#mode_selector").change(function() {
		var $selectedMode = $(":selected", this);
		humane.log("Mode changed to " + $selectedMode.html());
		var mode = $selectedMode.val();
		if(mode == 'clike') {
			changeMode(editor, mode, 'text/x-csrc');
		} else if(mode == 'plsql') {
			changeMode(editor, mode, 'text/x-plsql');
		} else {
			changeMode(editor, mode);
		}
	});

	var podChanged = function(cm) {
		$.post('/pod2html', {"source": cm.getValue()}, function(html) {
			$('#pod-tab').html(html);
		});
	};

	var editor = CodeMirror.fromTextArea(document.getElementById(editorId), {
		lineNumbers: true,
		matchBrackets: true,
		tabSize: 4,
		indentUnit: 4,
		indentWithTabs: true,
		extraKeys: {
			"F1": function(cm) {
				displayHelp(cm);
			},
			"F5": function(cm) {
				$(".run_in_browser_button").click();
			}
		}
	});

	// Hook up with cursor activity event
	editor.on("cursorActivity", function(cm) {
			// Highlight active line
			cm.setLineClass(hlLine, null, null);
			hlLine = cm.setLineClass(cm.getCursor().line, null, "activeline");

			// Highlight selection matches
			cm.matchHighlight("CodeMirror-matchhighlight");

			// Show editor statistics
			showEditorStats(cm);
	});

	// Hook up with cursor activity event
	var timeoutId;
	editor.on("change", function(cm) {
			clearInterval(timeoutId);
			timeoutId = setTimeout(function() {
				podChanged(cm);
				perl_critic();
				pod_check();
			}, 250);
	});

	// Run these when we exit this one
	setTimeout(function() {
		// Editor is by default Perl
		changeMode(editor, 'perl');

		// focus!
		editor.focus();

		// Show editor stats at startup
		showEditorStats(editor);

		// Update POD tab at startup
		podChanged(editor);

		// Run Perl::Critic at startup
		perl_critic();

		// Run Pod::Checker at startup
		pod_check();
	}, 0);

	// Highlight active line
	var hlLine = editor.setLineClass(0, "activeline");

	$(".results").hide();

	$(".run_in_browser_button").click(function() {
		// Load Perlito - Perl 5 to JS Compiler
		humane.log("Loading Perlito 5");
		$.ajax({
			url: 'assets/perlito/perlito5.min.js',
			dataType: "script",
			cache: true,
			success: function() {
				runOnPerlito(editor.getValue());
				humane.log("Run in browser finished");
			}
		});
	});

	var $output = $("#output");

	$(".perl_tidy_button").click(function() {
		$.post('/perl_tidy', {"source": editor.getValue()}, function(data) {
			if(data.error == '') {
				editor.setValue(data.source);
				humane.log("Your code is tidied!");
			} else {
				$output.val('Error:\n' + data.error);
				humane.log("Perl::Tidy failed! Please check output");
			}
		});
	});

	var perlCriticWidgets = [];

	var perl_critic = function() {

		$.post('/perl_critic', {"source": editor.getValue(), "severity": $(':selected', '#critic_severity_selector').val()}, function(violations) {
			for (var i = 0; i < perlCriticWidgets.length; i++) {
				editor.removeLineWidget(perlCriticWidgets[i].widget);
			}
			perlCriticWidgets.length = 0;

			if(violations.length > 0) {
				for(var i = 0; i < violations.length; i++) {
					var violation = violations[i];
					var description = violation.description;

					// Add Perl::Critic violation under the editor line
					var msg = $('<div class="farabi-error">' +
						'<span class="farabi-error-icon">!!</span>' + 
						description + '</div>')
						.appendTo(document)
						.get(0);
					perlCriticWidgets.push({
						violation: violation,
						widget:    editor.addLineWidget(violation.line_number - 1, msg, {coverGutter: true, noHScroll: true}),
						node:      msg
					});
				}

				$('.farabi-error').click(function() {
					
					var violation;
					for(var i = 0; i < perlCriticWidgets.length; i++) {
						if(perlCriticWidgets[i].node == this) {
							violation  = perlCriticWidgets[i].violation;
							break;
						}
					} 
					
					if(!violation) {
						return;
					}

					var policy = violation.policy;
					var html = '<strong>Explanation:</strong><br/>' + violation.explanation;
					html += '<br/><strong>Policy:</strong><br/><a target="_blank" href="https://metacpan.org/module/' + policy + '">' + policy + '</a>';
					html += '<br/><strong>Diagnositcs:</strong><br/>' + violation.diagnostics.replace(/\n/g, '<br/>');

					$('#myModalLabel', $perl_critic_dialog).html(
						violation.description
					);
					$('.modal-body', $perl_critic_dialog).html(html);
					$perl_critic_dialog.modal("show");
				});
			}
		});
	};

	$("#critic_severity_selector").change(function() {
		perl_critic();
	});

	var perlCheckerWidgets = [];

	var pod_check = function() {
		$.post('/pod_check', {"source": editor.getValue()}, function(problems) {
			for (var i = 0; i < perlCheckerWidgets.length; i++) {
				editor.removeLineWidget(perlCheckerWidgets[i].widget);
			}
			perlCheckerWidgets.length = 0;

			if(problems.length > 0) {
				for(var i = 0; i < problems.length; i++) {
					var problem = problems[i];

					// Add Pod::Checker warnings/error under the editor line
					var msg = $('<div class="farabi-error">' +
						'<span class="farabi-error-icon">!!</span>' + 
						problem.message + '</div>')
						.appendTo(document)
						.get(0);
					perlCheckerWidgets.push({
						widget:    editor.addLineWidget(problem.line - 1, msg, {coverGutter: true, noHScroll: true}),
					});
				}
			}
		});
	};

	$('.about_button').click(function() {
		if(typeof p5pkg == 'undefined') {
			humane.log("Loading Perlito 5");
			$.ajax({
				url: 'assets/perlito/perlito5.min.js',
				dataType: "script",
				cache: true,
				success: function() {
					$('#perlito-version').html(  p5pkg[ "main" ][ "v_]"]);
					humane.log("Perlito 5 loaded");
				}
			});
		} else {
			$('#perlito-version').html(p5pkg[ "main" ][ "v_]" ]);
		}
		$('#jquery-version').html($().jquery);
		$about_dialog.modal("show");
	});

	$('.help_button').click(function() {
		$help_dialog.modal("show");
	});

	var onCloseFocusEditor = function () {
		editor.focus();
	};
	$perl_critic_dialog.hide().on('hidden', onCloseFocusEditor);
	$about_dialog.hide().on('hidden', onCloseFocusEditor);
	$help_dialog.hide().on('hidden', onCloseFocusEditor);

	$(".topic").typeahead({
		source : function(query, process) {
			$.ajax({
				type: 'POST',
				url: '/typeahead',
				async: false,
				data: {'query': query},
				success: function(matches) {
					process(matches);
				},
				dataType: 'json',
			});
		}
	}).change(function() {
		_displayHelp($(this).val(), true);
	});;

	$("#line_numbers_checkbox").change(function() {
		editor.setOption('lineNumbers', $(this).is(':checked'));
	});

	$("#tab_size").change(function() {
		var tabSize = $(this).val();
		if($.isNumeric(tabSize)) {
			$(this).parent().parent().removeClass("error");
			editor.setOption('tabSize', tabSize);
		} else {
			$(this).parent().parent().addClass("error");
		}
	});
}

function runOnPerlito(source) {

	var $output = $('#output');

	// CORE.print prints to output tab
	p5pkg.CORE.print = function(List__) {
		var i;
		for (i = 0; i < List__.length; i++) {
			$output.val( $output.val() + p5str(List__[i]));
		}
		return true;
	};

	// CORE.warn print to output tab
	p5pkg.CORE.warn = function(List__) {
		var i;
		List__.push("\n");
		for (i = 0; i < List__.length; i++) {
			$output.val( $output.val() + p5str(List__[i]));
		}
		return true;
	};

	// Define version, strict and warnings
	p5pkg["main"]["v_^O"] = "browser";
	p5pkg["main"]["Hash_INC"]["Perlito5/strict.pm"] = "Perlito5/strict.pm";
	p5pkg["main"]["Hash_INC"]["Perlito5/warnings.pm"] = "Perlito5/warnings.pm";

	p5make_sub("Perlito5::IO", "slurp", function(filename) {
		console.error('IO.slurp "' + filename + '"');
		return 1;
	});

	p5is_file = function(filename) {
		console.error('p5is_file "' + filename + '"');
		return 1;
	}

	// Clear up output
	$output.val('');

	try {
		// Compile Perl 5 source code into JavaScript
		var start = $.now();
		var js_source = p5pkg["Perlito5"].compile_p5_to_js([source]);

		// Run JavaScript inside your browser
		start = $.now();;
		eval(js_source);

	}
	catch(err) {
		// Populate error and show error in output tab
		$output.val("Error:\n" + err + "\nCompilation aborted.\n");
	}

	// Show output tab
	$('a[href="#output-tab"]').tab('show');

}

// Start Farabi when the document loads
$(function() {
	startFarabi("editor");
});


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.016.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Modern Perl IDE",
  "AUTHOR" => "Ahmad M. Zawawi <ahmad.zawawi\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Farabi",
  "EXE_FILES" => [
    "bin/farabi"
  ],
  "LICENSE" => "perl",
  "NAME" => "Farabi",
  "PREREQ_PM" => {
    "Capture::Tiny" => "0.21",
    "DBD::SQLite" => "1.37",
    "DBIx::Simple" => "1.35",
    "File::Find::Rule" => "0.33",
    "File::HomeDir" => "1.00",
    "File::Which" => "1.09",
    "IO::String" => "1.08",
    "IPC::Run" => "0.92",
    "Mojolicious" => "4.63",
    "Parse::ErrorString::Perl" => "0.18",
    "Path::Tiny" => "0.013",
    "Text::Markdown" => "1.000031"
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::EOL" => "1.5",
    "Test::MinimumVersion" => "0.101080",
    "Test::More" => "0.98"
  },
  "VERSION" => "0.46",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Capture::Tiny" => "0.21",
  "DBD::SQLite" => "1.37",
  "DBIx::Simple" => "1.35",
  "File::Find::Rule" => "0.33",
  "File::HomeDir" => "1.00",
  "File::Spec" => 0,
  "File::Which" => "1.09",
  "IO::Handle" => 0,
  "IO::String" => "1.08",
  "IPC::Open3" => 0,
  "IPC::Run" => "0.92",
  "Mojolicious" => "4.63",
  "Parse::ErrorString::Perl" => "0.18",
  "Path::Tiny" => "0.013",
  "Test::EOL" => "1.5",
  "Test::MinimumVersion" => "0.101080",
  "Test::More" => "0.98",
  "Text::Markdown" => "1.000031"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



